<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Using crumblr in practice • crumblr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous">
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Using crumblr in practice">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">crumblr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.22</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/crumblr.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/crumblr_theory.html">Normal approximation vs. empirical simulation</a></li>
    <li><a class="dropdown-item" href="../articles/integration.html">Integration with dreamlet / SingleCellExperiment</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/DiseaseNeurogenomics/crumblr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Using crumblr in practice</h1>
                        <h4 data-toc-skip class="author">Developed by <a href="http://gabrielhoffman.github.io/" class="external-link">Gabriel Hoffman</a>
</h4>
            
            <h4 data-toc-skip class="date">Run on 2025-08-14
20:44:44.222783</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/DiseaseNeurogenomics/crumblr/blob/HEAD/vignettes/crumblr.Rmd" class="external-link"><code>vignettes/crumblr.Rmd</code></a></small>
      <div class="d-none name"><code>crumblr.Rmd</code></div>
    </div>

    
    
<style>
body {
text-align: justify}
</style>
<div class="section level3">
<h3 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h3>
<p>Changes in cell type composition play an important role in health and
disease. Recent advances in single cell technology have enabled
measurement of cell type composition at increasing cell lineage
resolution across large cohorts of individuals. Yet this raises new
challenges for statistical analysis of these compositional data to
identify changes associated with a phenotype. We introduce
<code>crumblr</code>, a scalable statistical method for analyzing count
ratio data using precision-weighted linear models incorporating random
effects for complex study designs. Uniquely, <code>crumblr</code>
performs tests of association at multiple levels of the cell lineage
hierarchy using multivariate regression to increase power over tests of
a single cell component. In simulations, <code>crumblr</code> increases
power compared to existing methods, while controlling the false positive
rate.</p>
<p>The <code>crumblr</code> package integrates with the <a href="https://www.bioconductor.org/packages/variancePartition/" class="external-link"><code>variancePartition</code></a>
and <a href="https://www.bioconductor.org/packages/dreamlet/" class="external-link"><code>dreamlet</code></a>
packages in the Bioconductor ecosystem.</p>
<p>Here we consider counts for 8 cell types from quantified using single
cell RNA-seq data from unstimulated and interferon β stimulated PBMCs
from 8 subjects <a href="https://www.nature.com/articles/nbt.4042" class="external-link">(Kang, et al.,
2018)</a>.</p>
<p>The functions here incorporate the precision weights:</p>
<ul>
<li><code><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/fitExtractVarPartModel-method.html" class="external-link">variancePartition::fitExtractVarPartModel()</a></code></li>
<li><code><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/dream-method.html" class="external-link">variancePartition::dream()</a></code></li>
<li><code><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">limma::lmFit()</a></code></li>
</ul>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install this package, start R and enter:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1) Make sure Bioconductor is installed</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># 2) Install crumblr and dependencies:</span></span>
<span><span class="co"># From Bioconductor</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"crumblr"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="analysis-workflow">Analysis workflow<a class="anchor" aria-label="anchor" href="#analysis-workflow"></a>
</h2>
<div class="section level3">
<h3 id="process-data">Process data<a class="anchor" aria-label="anchor" href="#process-data"></a>
</h3>
<p>Here we evaluate whether the observed cell proportions change in
response to interferon β. Given the results here, we cannot reject the
null hypothesis that interferon β does not affect the cell type
proportions.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://DiseaseNeurogenomics.github.io/crumblr" class="external-link">crumblr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load cell counts, clustering and metadata</span></span>
<span><span class="co"># from Kang, et al. (2018) https://doi.org/10.1038/nbt.4042</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">IFNCellCounts</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Apply crumblr transformation</span></span>
<span><span class="co"># cobj is an EList object compatable with limma workflow</span></span>
<span><span class="co"># cobj$E stores transformed values</span></span>
<span><span class="co"># cobj$weights stores precision weights</span></span>
<span><span class="co">#    corresponding to the regularized inverse variance</span></span>
<span><span class="va">cobj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/crumblr.html">crumblr</a></span><span class="op">(</span><span class="va">df_cellCounts</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="variance-partitioning">Variance partitioning<a class="anchor" aria-label="anchor" href="#variance-partitioning"></a>
</h3>
<p>Decomposing the variance illustrates that more variation is explained
by subject than stimulation status.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/variancePartition" class="external-link">variancePartition</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Partition variance into components for Subject (i.e. ind)</span></span>
<span><span class="co">#   and stimulation status, and residual variation</span></span>
<span><span class="va">form</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">ind</span><span class="op">)</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">StimStatus</span><span class="op">)</span></span>
<span><span class="va">vp</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/fitExtractVarPartModel-method.html" class="external-link">fitExtractVarPartModel</a></span><span class="op">(</span><span class="va">cobj</span>, <span class="va">form</span>, <span class="va">info</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot variance fractions</span></span>
<span><span class="va">fig.vp</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/plotPercentBars-method.html" class="external-link">plotPercentBars</a></span><span class="op">(</span><span class="va">vp</span><span class="op">)</span></span>
<span><span class="va">fig.vp</span></span></code></pre></div>
<p><img src="crumblr_files/figure-html/vp-1.png" width="576"></p>
</div>
<div class="section level3">
<h3 id="pca">PCA<a class="anchor" aria-label="anchor" href="#pca"></a>
</h3>
<p>Performing PCA on the transformed cell counts indicates that the
samples cluster based on subject rather than stimulation status. Here,
we use <code><a href="../reference/standardize-method.html">standardize()</a></code> so that each observation has
approximately equal variance (i.e. homoscedasticity) by dividing the CLR
transformed frequencies by their estimated sampling standard deviation.
Transforming the data to be approximately homoscedastic has been <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8" class="external-link">shown</a>
to improve performance of PCA.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform PCA</span></span>
<span><span class="co"># use crumblr::standardize() to get values with</span></span>
<span><span class="co"># approximately equal sampling variance,</span></span>
<span><span class="co"># which is a key property for downstream PCA and clustering analysis.</span></span>
<span><span class="va">pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="../reference/standardize-method.html">standardize</a></span><span class="op">(</span><span class="va">cobj</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># merge with metadata</span></span>
<span><span class="va">df_pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">pca</span><span class="op">$</span><span class="va">x</span>, <span class="va">info</span>, by <span class="op">=</span> <span class="st">"row.names"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot PCA</span></span>
<span><span class="co">#   color by Subject</span></span>
<span><span class="co">#   shape by Stimulated vs unstimulated</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df_pca</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">PC1</span>, <span class="va">PC2</span>, color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">ind</span><span class="op">)</span>, shape <span class="op">=</span> <span class="va">StimStatus</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_colour_discrete.html" class="external-link">scale_color_discrete</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Subject"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"PC1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"PC2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="crumblr_files/figure-html/pca-1.png" width="480"></p>
</div>
<div class="section level3">
<h3 id="hierarchical-clustering">Hierarchical clustering<a class="anchor" aria-label="anchor" href="#hierarchical-clustering"></a>
</h3>
<p>The samples from the same subject also cluster together.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html" class="external-link">heatmap</a></span><span class="op">(</span><span class="va">cobj</span><span class="op">$</span><span class="va">E</span><span class="op">)</span></span></code></pre></div>
<p><img src="crumblr_files/figure-html/hclust-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="differential-testing">Differential testing<a class="anchor" aria-label="anchor" href="#differential-testing"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use variancePartition workflow to analyze each cell type</span></span>
<span><span class="co"># Perform regression on each cell type separately</span></span>
<span><span class="co">#  then use eBayes to shrink residual variance</span></span>
<span><span class="co"># Also compatible with limma::lmFit()</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/dream-method.html" class="external-link">dream</a></span><span class="op">(</span><span class="va">cobj</span>, <span class="op">~</span> <span class="va">StimStatus</span> <span class="op">+</span> <span class="va">ind</span>, <span class="va">info</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract results for each cell type</span></span>
<span><span class="fu">topTable</span><span class="op">(</span><span class="va">fit</span>, coef <span class="op">=</span> <span class="st">"StimStatusstim"</span>, number <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                         logFC    AveExpr          t     P.Value  adj.P.Val         B</span></span>
<span><span class="co">## CD8 T cells       -0.25085170  0.0857175 -4.0787416 0.002436375 0.01949100 -1.279815</span></span>
<span><span class="co">## Dendritic cells    0.37386979 -2.1849234  3.1619195 0.010692544 0.02738587 -2.638507</span></span>
<span><span class="co">## CD14+ Monocytes   -0.10525402  1.2698117 -3.1226341 0.011413912 0.02738587 -2.709377</span></span>
<span><span class="co">## B cells           -0.10478652  0.5516882 -3.0134349 0.013692935 0.02738587 -2.940542</span></span>
<span><span class="co">## CD4 T cells       -0.07840101  2.0201947 -2.2318104 0.050869691 0.08139151 -4.128069</span></span>
<span><span class="co">## FCGR3A+ Monocytes  0.07425165 -0.2567492  1.6647681 0.128337022 0.17111603 -4.935304</span></span>
<span><span class="co">## NK cells           0.10270672  0.3797777  1.5181860 0.161321761 0.18436773 -5.247806</span></span>
<span><span class="co">## Megakaryocytes     0.01377768 -1.8655172  0.1555131 0.879651456 0.87965146 -6.198336</span></span></code></pre>
<div class="section level4">
<h4 id="multivariate-testing-along-a-tree">Multivariate testing along a tree<a class="anchor" aria-label="anchor" href="#multivariate-testing-along-a-tree"></a>
</h4>
<p>We can gain power by jointly testing multiple cell types using a
multivariate statistical model, instead of testing one cell type at a
time. Here we construct a hierarchical clustering between cell types
based on gene expression from pseudobulk, and perform a multivariate
test for each internal node of the tree based on its leaf nodes. The
results for the leaves are the same as from <code>topTable()</code>
above. At each internal node <code><a href="../reference/treeTest.html">treeTest()</a></code> performs a fixed
effects meta-analysis of the coefficients of the leaves while modeling
the covariance between coefficient estimates. In the backend, this is
implemented using <code><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/mvTest-method.html" class="external-link">variancePartition::mvTest()</a></code> and <a href="https://cran.r-project.org/package=remaCor" class="external-link">remaCor</a>
package.</p>
<p>Here the hierarchical clustering, <code>hcl</code>, is precomputed
from pseudobulk gene expression using
<code>buildClusterTreeFromPB()</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform multivariate test across the hierarchy</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/treeTest.html">treeTest</a></span><span class="op">(</span><span class="va">fit</span>, <span class="va">cobj</span>, <span class="va">hcl</span>, coef <span class="op">=</span> <span class="st">"StimStatusstim"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot hierarchy and testing results</span></span>
<span><span class="fu"><a href="../reference/plotTreeTest.html">plotTreeTest</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<p><img src="crumblr_files/figure-html/treeTest-1.png" width="700"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot hierarchy and regression coefficients</span></span>
<span><span class="fu"><a href="../reference/plotTreeTestBeta.html">plotTreeTestBeta</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<p><img src="crumblr_files/figure-html/treeTest-2.png" width="700"></p>
<div class="section level5">
<h5 id="combined-plotting">Combined plotting<a class="anchor" aria-label="anchor" href="#combined-plotting"></a>
</h5>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotTreeTestBeta.html">plotTreeTestBeta</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span>, legend.box <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span> <span class="op">|</span></span>
<span>  <span class="fu"><a href="../reference/plotForest-methods.html">plotForest</a></span><span class="op">(</span><span class="va">res</span>, hide <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">|</span></span>
<span>  <span class="va">fig.vp</span></span></code></pre></div>
<p><img src="crumblr_files/figure-html/combined-1.png" width="1152"></p>
<p><br></p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="hierarchical-clustering-1">Hierarchical clustering<a class="anchor" aria-label="anchor" href="#hierarchical-clustering-1"></a>
</h3>
<p>The hierarchical clustering used by <code><a href="../reference/treeTest.html">treeTest()</a></code> can be
computed a number of ways, depending on the available data and
biological question. For example, see <a href="integration.html">Article</a> for details about how hierarchical
clustering was run in this dataset.</p>
<p>In general, hierarchical clustering can be computed from</p>
<ul>
<li>pseudobulked single cell gene expression</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hcl</span> <span class="op">&lt;-</span> <span class="fu">buildClusterTreeFromPB</span><span class="op">(</span><span class="va">pb</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>cell type frequencies</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># correlation matrix between all cell types</span></span>
<span><span class="va">C</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="../reference/standardize-method.html">standardize</a></span><span class="op">(</span><span class="va">cobj</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># convert to distance</span></span>
<span><span class="va">dm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">as.dist</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">C</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># eval hierarchical clustering</span></span>
<span><span class="va">hcl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">hclust</a></span><span class="op">(</span><span class="va">dm</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>
<a href="https://en.wikipedia.org/wiki/Newick_format" class="external-link">Newick
formated</a> tree computed from external data</li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Make sure packages are installed</span></span>
<span><span class="co"># BiocManager::install(c("ctc", "ape", "phylogram"))</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/emmanuelparadis/ape" class="external-link">ape</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://antoinelucas.free.fr/ctc" class="external-link">ctc</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://docs.ropensci.org/phylogram" class="external-link">phylogram</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Write tree to file, </span></span>
<span><span class="co"># edit manually</span></span>
<span><span class="co"># then read back into R</span></span>
<span><span class="co"># </span></span>
<span></span>
<span><span class="co"># Specify tree as text in Newick format</span></span>
<span><span class="va">txt</span> <span class="op">=</span> <span class="st">"((CD14+ Monocytes,(B cells,(Dendritic cells,Megakaryocytes))),(CD8 T cells,(NK cells,(CD4 T cells,FCGR3A+ Monocytes))));"</span></span>
<span></span>
<span><span class="co"># read from text</span></span>
<span><span class="va">hcl_from_txt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span>text <span class="op">=</span> <span class="va">txt</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                  <span class="va">as.dendrogram.phylo</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                  <span class="va">as.hclust</span></span>
<span></span>
<span><span class="co"># Alternatively, </span></span>
<span><span class="co"># write existing tree to file</span></span>
<span><span class="co"># and edit manaully</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/write.html" class="external-link">write</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ctc/man/hc2Newick.html" class="external-link">hc2Newick</a></span><span class="op">(</span><span class="va">hcl</span><span class="op">)</span>,file<span class="op">=</span><span class="st">'hcl.nwk'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hcl_from_txt2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">'hcl.nwk'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                  <span class="va">as.dendrogram.phylo</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                  <span class="va">as.hclust</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="considerations">Considerations<a class="anchor" aria-label="anchor" href="#considerations"></a>
</h2>
<div class="section level3">
<h3 id="computational-scaling">Computational scaling<a class="anchor" aria-label="anchor" href="#computational-scaling"></a>
</h3>
<p>The <code><a href="../reference/crumblr.html">crumblr()</a></code> workflow is very fast, especially compared
to more demanding differential expression analysies. Applying
<code><a href="../reference/crumblr.html">crumblr()</a></code> requires &lt;1 sec, even for very large datsets.
Differential testing with <code><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/dream-method.html" class="external-link">dream()</a></code> takes &lt; 10 seconds for
fixed effects models and &lt; 30 seconds for mixed models with typical
sample sizes and number of cell types. Running <code><a href="../reference/treeTest.html">treeTest()</a></code>
can be a little more demanding, but should finish in &lt; 30 seconds
with 20 cell types.</p>
</div>
<div class="section level3">
<h3 id="complex-study-designs">Complex study designs<a class="anchor" aria-label="anchor" href="#complex-study-designs"></a>
</h3>
<p>The <code><a href="../reference/crumblr.html">crumblr()</a></code> workflow can handle complex study designs
with repeated measures or multiple random effects. <code><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/dream-method.html" class="external-link">dream()</a></code>
uses <code><a href="https://rdrr.io/pkg/lme4/man/lmer.html" class="external-link">lme4::lmer()</a></code> in the backend to fit weighted linear
mixed models. Considerations about defining the regression model are
described in documentation to <a href="https://diseaseneurogenomics.github.io/variancePartition" class="external-link"><code>variancePartition</code></a>
or this <a href="https://people.math.ethz.ch/~maechler/MEMo-pages/lMMwR.pdf" class="external-link">book</a>
by the author of <code>lme4</code>.</p>
</div>
<div class="section level3">
<h3 id="tuning-parameters">Tuning parameters<a class="anchor" aria-label="anchor" href="#tuning-parameters"></a>
</h3>
<p><code><a href="../reference/crumblr.html">crumblr()</a></code> uses two tuning parameters accessable to the
user. These are fixed to default values in all simulations and data
analysis. <u>The user is strongly recommended to keep these dfault
values</u>.</p>
<ul>
<li><p>In order to deal with zero counts, <code><a href="../reference/crumblr.html">crumblr()</a></code> uses a
pseudocount (default: 0.5) added to all observed counts.</p></li>
<li><p>For real data, the asymptotic variance formula can give weights
that vary substantially across samples and give very high weights for a
subset of samples. In order to address this, we regularize the weights
to reduce the variation in the weights to have a maximum ratio (default
of 5) between the maximum and specified quantile value (default of
5%).</p></li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<details><pre><code><span><span class="co">## R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin23.6.0</span></span>
<span><span class="co">## Running under: macOS Sonoma 14.7.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS/LAPACK: /opt/homebrew/Cellar/openblas/0.3.30/lib/libopenblasp-r0.3.30.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/New_York</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] muscat_1.22.0            crumblr_0.99.22          variancePartition_1.37.4</span></span>
<span><span class="co">## [4] BiocParallel_1.42.1      limma_3.64.3             ggplot2_3.5.2           </span></span>
<span><span class="co">## [7] BiocStyle_2.36.0        </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] splines_4.5.1               bitops_1.0-9                ggplotify_0.1.2            </span></span>
<span><span class="co">##   [4] tibble_3.3.0                lifecycle_1.0.4             Rdpack_2.6.4               </span></span>
<span><span class="co">##   [7] edgeR_4.6.3                 doParallel_1.0.17           globals_0.18.0             </span></span>
<span><span class="co">##  [10] lattice_0.22-7              MASS_7.3-65                 backports_1.5.0            </span></span>
<span><span class="co">##  [13] magrittr_2.0.3              sass_0.4.10                 rmarkdown_2.29             </span></span>
<span><span class="co">##  [16] jquerylib_0.1.4             yaml_2.3.10                 sctransform_0.4.2          </span></span>
<span><span class="co">##  [19] minqa_1.2.8                 RColorBrewer_1.1-3          abind_1.4-8                </span></span>
<span><span class="co">##  [22] EnvStats_3.1.0              glmmTMB_1.1.11              GenomicRanges_1.60.0       </span></span>
<span><span class="co">##  [25] purrr_1.1.0                 BiocGenerics_0.54.0         yulab.utils_0.2.0          </span></span>
<span><span class="co">##  [28] circlize_0.4.16             GenomeInfoDbData_1.2.14     IRanges_2.42.0             </span></span>
<span><span class="co">##  [31] S4Vectors_0.46.0            ggrepel_0.9.6               pbkrtest_0.5.5             </span></span>
<span><span class="co">##  [34] irlba_2.3.5.1               listenv_0.9.1               tidytree_0.4.6             </span></span>
<span><span class="co">##  [37] parallelly_1.45.1           pkgdown_2.1.3               codetools_0.2-20           </span></span>
<span><span class="co">##  [40] DelayedArray_0.34.1         scuttle_1.18.0              tidyselect_1.2.1           </span></span>
<span><span class="co">##  [43] shape_1.4.6.1               aplot_0.2.8                 UCSC.utils_1.4.0           </span></span>
<span><span class="co">##  [46] farver_2.1.2                lme4_1.1-37                 ScaledMatrix_1.16.0        </span></span>
<span><span class="co">##  [49] viridis_0.6.5               matrixStats_1.5.0           stats4_4.5.1               </span></span>
<span><span class="co">##  [52] jsonlite_2.0.0              GetoptLong_1.0.5            BiocNeighbors_2.2.0        </span></span>
<span><span class="co">##  [55] scater_1.36.0               iterators_1.0.14            systemfonts_1.2.3          </span></span>
<span><span class="co">##  [58] foreach_1.5.2               tools_4.5.1                 progress_1.2.3             </span></span>
<span><span class="co">##  [61] treeio_1.32.0               ragg_1.4.0                  Rcpp_1.1.0                 </span></span>
<span><span class="co">##  [64] blme_1.0-6                  glue_1.8.0                  gridExtra_2.3              </span></span>
<span><span class="co">##  [67] SparseArray_1.8.1           xfun_0.52                   mgcv_1.9-3                 </span></span>
<span><span class="co">##  [70] DESeq2_1.48.1               MatrixGenerics_1.20.0       GenomeInfoDb_1.44.1        </span></span>
<span><span class="co">##  [73] dplyr_1.1.4                 withr_3.0.2                 numDeriv_2016.8-1.1        </span></span>
<span><span class="co">##  [76] BiocManager_1.30.26         fastmap_1.2.0               boot_1.3-31                </span></span>
<span><span class="co">##  [79] caTools_1.18.3              digest_0.6.37               rsvd_1.0.5                 </span></span>
<span><span class="co">##  [82] R6_2.6.1                    gridGraphics_0.5-1          textshaping_1.0.1          </span></span>
<span><span class="co">##  [85] colorspace_2.1-1            gtools_3.9.5                RhpcBLASctl_0.23-42        </span></span>
<span><span class="co">##  [88] tidyr_1.3.1                 generics_0.1.4              data.table_1.17.8          </span></span>
<span><span class="co">##  [91] corpcor_1.6.10              prettyunits_1.2.0           httr_1.4.7                 </span></span>
<span><span class="co">##  [94] htmlwidgets_1.6.4           S4Arrays_1.8.1              pkgconfig_2.0.3            </span></span>
<span><span class="co">##  [97] gtable_0.3.6                ComplexHeatmap_2.24.1       SingleCellExperiment_1.30.1</span></span>
<span><span class="co">## [100] XVector_0.48.0              remaCor_0.0.18              htmltools_0.5.8.1          </span></span>
<span><span class="co">## [103] bookdown_0.43               TMB_1.9.17                  zigg_0.0.2                 </span></span>
<span><span class="co">## [106] clue_0.3-66                 scales_1.4.0                Biobase_2.68.0             </span></span>
<span><span class="co">## [109] png_0.1-8                   fANCOVA_0.6-1               reformulas_0.4.1           </span></span>
<span><span class="co">## [112] ggfun_0.2.0                 knitr_1.50                  reshape2_1.4.4             </span></span>
<span><span class="co">## [115] rjson_0.2.23                nlme_3.1-168                nloptr_2.2.1               </span></span>
<span><span class="co">## [118] cachem_1.1.0                GlobalOptions_0.1.2         stringr_1.5.1              </span></span>
<span><span class="co">## [121] KernSmooth_2.23-26          parallel_4.5.1              vipor_0.4.7                </span></span>
<span><span class="co">## [124] desc_1.4.3                  pillar_1.11.0               grid_4.5.1                 </span></span>
<span><span class="co">## [127] vctrs_0.6.5                 gplots_3.2.0                BiocSingular_1.24.0        </span></span>
<span><span class="co">## [130] beachmat_2.24.0             cluster_2.1.8.1             beeswarm_0.4.0             </span></span>
<span><span class="co">## [133] evaluate_1.0.4              mvtnorm_1.3-3               cli_3.6.5                  </span></span>
<span><span class="co">## [136] locfit_1.5-9.12             compiler_4.5.1              rlang_1.1.6                </span></span>
<span><span class="co">## [139] crayon_1.5.3                future.apply_1.20.0         labeling_0.4.3             </span></span>
<span><span class="co">## [142] plyr_1.8.9                  fs_1.6.6                    ggbeeswarm_0.7.2           </span></span>
<span><span class="co">## [145] stringi_1.8.7               viridisLite_0.4.2           lmerTest_3.1-3             </span></span>
<span><span class="co">## [148] lazyeval_0.2.2              aod_1.3.3                   Matrix_1.7-3               </span></span>
<span><span class="co">## [151] hms_1.1.3                   patchwork_1.3.1             future_1.67.0              </span></span>
<span><span class="co">## [154] statmod_1.5.0               SummarizedExperiment_1.38.1 rbibutils_2.3              </span></span>
<span><span class="co">## [157] Rfast_2.1.5.1               broom_1.0.9                 RcppParallel_5.1.10.9000   </span></span>
<span><span class="co">## [160] bslib_0.9.0                 ggtree_3.16.3               ape_5.8-1</span></span></code></pre>
</details>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
