<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="crumblr">
<title>Multivariate testing along a tree • crumblr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Multivariate testing along a tree">
<meta property="og:description" content="crumblr">
<meta property="og:image" content="http://gabrielhoffman.github.io/crumblr/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">crumblr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/crumblr.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/crumblr_theory.html">Normal approximation vs. empirical simulation</a>
    <a class="dropdown-item" href="../articles/crumblr_treeTest.html">Multivariate testing along a tree</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/GabrielHoffman/crumblr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Multivariate testing along a tree</h1>
                        <h4 data-toc-skip class="author">Developed by <a href="http://gabrielhoffman.github.io/" class="external-link">Gabriel Hoffman</a>
</h4>
            
            <h4 data-toc-skip class="date">Run on 2022-12-07
20:21:55</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/GabrielHoffman/crumblr/blob/HEAD/vignettes/crumblr_treeTest.Rmd" class="external-link"><code>vignettes/crumblr_treeTest.Rmd</code></a></small>
      <div class="d-none name"><code>crumblr_treeTest.Rmd</code></div>
    </div>

    
    
<!---
cd /Users/gabrielhoffman/workspace/repos/crumblr/vignettes
R

rmarkdown::render('crumblr_treeTest.Rmd')
--->
<style>
body {
text-align: justify}
</style>
<div class="section level2">
<h2 id="load-and-process-single-cell-data">Load and process single cell data<a class="anchor" aria-label="anchor" href="#load-and-process-single-cell-data"></a>
</h2>
<p>Here we perform analysis of PBMCs from 8 individuals stimulated with
interferon-β <a href="https://www.nature.com/articles/nbt.4042" class="external-link">Kang, et
al, 2018, Nature Biotech</a>. We perform standard processing with <a href="https://gabrielhoffman.github.io/dreamlet/index.html" class="external-link">dreamlet</a>
to compute pseudobulk before applying <code>crumblr</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://gabrielhoffman.github.io/dreamlet" class="external-link">dreamlet</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HelenaLC/muscat" class="external-link">muscat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ExperimentHub</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Download data, specifying EH2259 for the Kang, et al study</span></span>
<span><span class="va">eh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ExperimentHub/man/ExperimentHub-class.html" class="external-link">ExperimentHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH2259"</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># only keep singlet cells with sufficient reads</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="op">]</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">multiplets</span> <span class="op">==</span> <span class="st">'singlet'</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># compute QC metrics</span></span>
<span><span class="va">qc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html" class="external-link">perCellQCMetrics</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># remove cells with few or many detected genes</span></span>
<span><span class="va">ol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span>metric <span class="op">=</span> <span class="va">qc</span><span class="op">$</span><span class="va">detected</span>, nmads <span class="op">=</span> <span class="fl">2</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="op">!</span><span class="va">ol</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># set variable indicating stimulated (stim) or control (ctrl)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">StimStatus</span> <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">stim</span></span>
<span></span>
<span><span class="co"># Since 'ind' is the individual and 'StimStatus' is the stimulus status,</span></span>
<span><span class="co"># create unique identifier for each sample</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">StimStatus</span>, <span class="va">sce</span><span class="op">$</span><span class="va">ind</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create pseudobulk data by specifying cluster_id and sample_id</span></span>
<span><span class="co"># Count data for each cell type is then stored in the `assay` field</span></span>
<span><span class="co"># assay: entry in assayNames(sce) storing raw counts</span></span>
<span><span class="co"># cluster_id: variable in colData(sce) indicating cell clusters</span></span>
<span><span class="co"># sample_id: variable in colData(sce) indicating sample id for aggregating cells</span></span>
<span><span class="va">pb</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://gabrielhoffman.github.io/dreamlet/reference/aggregateToPseudoBulk.html" class="external-link">aggregateToPseudoBulk</a></span><span class="op">(</span><span class="va">sce</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"counts"</span>,     </span>
<span>    cluster_id <span class="op">=</span> <span class="st">"cell"</span>,  </span>
<span>    sample_id <span class="op">=</span> <span class="st">"id"</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="standard-crumblr-analysis">Standard <code>crumblr</code> analysis<a class="anchor" aria-label="anchor" href="#standard-crumblr-analysis"></a>
</h2>
<p>Apply standard <code>crumblr</code> analysis to transform data and
apply a linear (mixed) model, followed by empirical Bayes moderated
statistics.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://gabrielhoffman.github.io/crumblr" class="external-link">crumblr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># crumblr transformation using</span></span>
<span><span class="co"># cellCounts() to extract counts from pseudobulk</span></span>
<span><span class="va">cobj</span> <span class="op">=</span> <span class="fu"><a href="../reference/crumblr.html">crumblr</a></span><span class="op">(</span><span class="fu"><a href="../reference/IFNCellCounts.html">cellCounts</a></span><span class="op">(</span><span class="va">pb</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Use dream workflow to analyze each cell separately</span></span>
<span><span class="va">fit</span> <span class="op">=</span> <span class="fu"><a href="http://gabrielhoffman.github.io/variancePartition/reference/dream-method.html" class="external-link">dream</a></span><span class="op">(</span><span class="va">cobj</span>, <span class="op">~</span> <span class="va">StimStatus</span> <span class="op">+</span> <span class="va">ind</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">pb</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extract top results</span></span>
<span><span class="fu">topTable</span><span class="op">(</span><span class="va">fit</span>, coef<span class="op">=</span><span class="st">"StimStatusstim"</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">##                         logFC    AveExpr          t   P.Value adj.P.Val         B</span></span>
<span><span class="co">## Dendritic cells    0.32516900 -2.1849234  1.0895217 0.2923251  0.823424 -4.590653</span></span>
<span><span class="co">## CD14+ Monocytes   -0.12437220  1.2698117 -0.9560159 0.3535191  0.823424 -4.685558</span></span>
<span><span class="co">## B cells           -0.09825871  0.5516882 -0.6779201 0.5076669  0.823424 -4.746255</span></span>
<span><span class="co">## NK cells           0.08494553  0.3797777  0.5814645 0.5691622  0.823424 -4.753355</span></span>
<span><span class="co">## CD4 T cells       -0.07763160  2.0201947 -0.4732415 0.6425378  0.823424 -4.900413</span></span>
<span><span class="co">## CD8 T cells       -0.16431725  0.0857175 -0.3892080 0.7023377  0.823424 -4.758201</span></span>
<span><span class="co">## FCGR3A+ Monocytes  0.06580754 -0.2567492  0.3460969 0.7338531  0.823424 -4.734622</span></span>
<span><span class="co">## Megakaryocytes     0.04681137 -1.8655172  0.2268978 0.8234240  0.823424 -4.635391</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="multivariate-testing-along-a-tree">Multivariate testing along a tree<a class="anchor" aria-label="anchor" href="#multivariate-testing-along-a-tree"></a>
</h2>
<p>We can gain power by jointly testing multiple cell types using a
multivariate statistical model, instead of testing one cell type at a
time. Here we construct a hierarchical clustering between cell types
based on gene expression from pseudobulk, and perform a multivariate
test for each internal node of the tree based on its leaf nodes. The
results for the leaves are the same as from <code>topTable()</code>
above. At each internal node <code><a href="../reference/treeTest.html">treeTest()</a></code> performs a fixed
effects meta-analysis of the coefficients of the leaves while modeling
the covariance between coefficient estimates. In the backend, this is
implemented using <code><a href="http://gabrielhoffman.github.io/variancePartition/reference/mvTest.html" class="external-link">variancePartition::mvTest()</a></code> and <a href="https://cran.r-project.org/package=remaCor" class="external-link">remaCor
package</a>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Hierarchical clustering of cell types</span></span>
<span><span class="co"># The tree is based on similarity between cell types</span></span>
<span><span class="co"># based on gene expression</span></span>
<span><span class="va">hcl</span> <span class="op">=</span> <span class="fu"><a href="http://gabrielhoffman.github.io/dreamlet/reference/buildClusterTreeFromPB.html" class="external-link">buildClusterTreeFromPB</a></span><span class="op">(</span><span class="va">pb</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform multivariate test across the hierarchy</span></span>
<span><span class="co"># Specify coefficient to test</span></span>
<span><span class="va">res</span> <span class="op">=</span> <span class="fu"><a href="../reference/treeTest.html">treeTest</a></span><span class="op">(</span> <span class="va">fit</span>, <span class="va">cobj</span>, <span class="va">hcl</span>, coef<span class="op">=</span><span class="st">"StimStatusstim"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot hierarchy and testing results</span></span>
<span><span class="co"># Adjust xlim() until text fits in window </span></span>
<span><span class="fu"><a href="../reference/plotTreeTest.html">plotTreeTest</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">7</span><span class="op">)</span></span></code></pre></div>
<p><img src="crumblr_treeTest_files/figure-html/treeTest-1.png" width="700"></p>
<p>In this example, no result passes the multiple testing cutoff.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
