<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Using crumblr in practice • crumblr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Using crumblr in practice"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">crumblr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.19</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/crumblr.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/crumblr_theory.html">Normal approximation vs. empirical simulation</a></li>
    <li><a class="dropdown-item" href="../articles/integration.html">Integration with dreamlet / SingleCellExperiment</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/DiseaseNeurogenomics/crumblr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Using crumblr in practice</h1>
                        <h4 data-toc-skip class="author">Developed by <a
href="http://gabrielhoffman.github.io/">Gabriel Hoffman</a></h4>
            
            <h4 data-toc-skip class="date">Run on 2025-02-17
21:01:43.120205</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/DiseaseNeurogenomics/crumblr/blob/HEAD/vignettes/crumblr.Rmd" class="external-link"><code>vignettes/crumblr.Rmd</code></a></small>
      <div class="d-none name"><code>crumblr.Rmd</code></div>
    </div>

    
    
<style>
body {
text-align: justify}
</style>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>Changes in cell type composition play an important role in health and
disease. Recent advances in single cell technology have enabled
measurement of cell type composition at increasing cell lineage
resolution across large cohorts of individuals. Yet this raises new
challenges for statistical analysis of these compositional data to
identify changes associated with a phenotype. We introduce
<code>crumblr</code>, a scalable statistical method for analyzing count
ratio data using precision-weighted linear models incorporating random
effects for complex study designs. Uniquely, <code>crumblr</code>
performs tests of association at multiple levels of the cell lineage
hierarchy using multivariate regression to increase power over tests of
a single cell component. In simulations, <code>crumblr</code> increases
power compared to existing methods, while controlling the false positive
rate.</p>
<p>The <code>crumblr</code> package integrates with the <a
href="https://www.bioconductor.org/packages/variancePartition/"><code>variancePartition</code></a>
and <a
href="https://www.bioconductor.org/packages/dreamlet/"><code>dreamlet</code></a>
packages in the Bioconductor ecosystem.</p>
<p>Here we consider counts for 8 cell types from quantified using single
cell RNA-seq data from unstimulated and interferon β stimulated PBMCs
from 8 subjects <a
href="https://www.nature.com/articles/nbt.4042">(Kang, et al.,
2018)</a>.</p>
<p>The functions here incorporate the precision weights:</p>
<ul>
<li><code>variancePartition::fitExtractVarPartModel()</code></li>
<li><code>variancePartition::dream()</code></li>
<li><code>limma::lmFit()</code></li>
</ul>
</div>
<div id="installation" class="section level1">
<h1>Installation</h1>
<p>To install this package, start R and enter:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># 1) Make sure Bioconductor is installed</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)) {</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>}</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="co"># 2) Install crumblr and dependencies:</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="co"># From Bioconductor</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;crumblr&quot;</span>)</span></code></pre></div>
<div id="process-data" class="section level2">
<h2>Process data</h2>
<p>Here we evaluate whether the observed cell proportions change in
response to interferon β. Given the results here, we cannot reject the
null hypothesis that interferon β does not affect the cell type
proportions.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">library</span>(crumblr)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="co"># Load cell counts, clustering and metadata</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co"># from Kang, et al. (2018) https://doi.org/10.1038/nbt.4042</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="fu">data</span>(IFNCellCounts)</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="co"># Apply crumblr transformation</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="co"># cobj is an EList object compatable with limma workflow</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="co"># cobj$E stores transformed values</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co"># cobj$weights stores precision weights</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="co">#    corresponding to the regularized inverse variance</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a>cobj <span class="ot">&lt;-</span> <span class="fu">crumblr</span>(df_cellCounts)</span></code></pre></div>
</div>
<div id="variance-partitioning" class="section level2">
<h2>Variance partitioning</h2>
<p>Decomposing the variance illustrates that more variation is explained
by subject than stimulation status.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">library</span>(variancePartition)</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="co"># Partition variance into components for Subject (i.e. ind)</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="co">#   and stimulation status, and residual variation</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>form <span class="ot">&lt;-</span> <span class="er">~</span> (<span class="dv">1</span> <span class="sc">|</span> ind) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> StimStatus)</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>vp <span class="ot">&lt;-</span> <span class="fu">fitExtractVarPartModel</span>(cobj, form, info)</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="co"># Plot variance fractions</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>fig.vp <span class="ot">&lt;-</span> <span class="fu">plotPercentBars</span>(vp)</span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>fig.vp</span></code></pre></div>
<p><img src="/Users/gabrielhoffman/workspace/repos/crumblr/docs/articles/crumblr_files/figure-html/vp-1.png" width="576" /></p>
</div>
<div id="pca" class="section level2">
<h2>PCA</h2>
<p>Performing PCA on the transformed cell counts indicates that the
samples cluster based on subject rather than stimulation status.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="co"># Perform PCA</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co"># use crumblr::standardize() to get values with</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co"># approximately equal sampling variance,</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="co"># which is a key property for downstream PCA and clustering analysis.</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>pca <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(<span class="fu">t</span>(<span class="fu">standardize</span>(cobj)))</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="co"># merge with metadata</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a>df_pca <span class="ot">&lt;-</span> <span class="fu">merge</span>(pca<span class="sc">$</span>x, info, <span class="at">by =</span> <span class="st">&quot;row.names&quot;</span>)</span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a><span class="co"># Plot PCA</span></span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a><span class="co">#   color by Subject</span></span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a><span class="co">#   shape by Stimulated vs unstimulated</span></span>
<span id="cb4-15"><a href="#cb4-15" tabindex="-1"></a><span class="fu">ggplot</span>(df_pca, <span class="fu">aes</span>(PC1, PC2, <span class="at">color =</span> <span class="fu">as.character</span>(ind), <span class="at">shape =</span> StimStatus)) <span class="sc">+</span></span>
<span id="cb4-16"><a href="#cb4-16" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb4-17"><a href="#cb4-17" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb4-18"><a href="#cb4-18" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">aspect.ratio =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb4-19"><a href="#cb4-19" tabindex="-1"></a>  <span class="fu">scale_color_discrete</span>(<span class="at">name =</span> <span class="st">&quot;Subject&quot;</span>) <span class="sc">+</span></span>
<span id="cb4-20"><a href="#cb4-20" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;PC1&quot;</span>) <span class="sc">+</span></span>
<span id="cb4-21"><a href="#cb4-21" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;PC2&quot;</span>)</span></code></pre></div>
<p><img src="/Users/gabrielhoffman/workspace/repos/crumblr/docs/articles/crumblr_files/figure-html/pca-1.png" width="480" /></p>
</div>
<div id="hierarchical-clustering" class="section level2">
<h2>Hierarchical clustering</h2>
<p>The samples from the same subject also cluster together.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">heatmap</span>(cobj<span class="sc">$</span>E)</span></code></pre></div>
<p><img src="/Users/gabrielhoffman/workspace/repos/crumblr/docs/articles/crumblr_files/figure-html/hclust-1.png" width="700" /></p>
</div>
<div id="differential-testing" class="section level2">
<h2>Differential testing</h2>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># Use variancePartition workflow to analyze each cell type</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="co"># Perform regression on each cell type separately</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="co">#  then use eBayes to shrink residual variance</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="co"># Also compatible with limma::lmFit()</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">dream</span>(cobj, <span class="sc">~</span> StimStatus <span class="sc">+</span> ind, info)</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">eBayes</span>(fit)</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="co"># Extract results for each cell type</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="fu">topTable</span>(fit, <span class="at">coef =</span> <span class="st">&quot;StimStatusstim&quot;</span>, <span class="at">number =</span> <span class="cn">Inf</span>)</span></code></pre></div>
<pre><code>##                         logFC    AveExpr          t     P.Value  adj.P.Val         B
## CD8 T cells       -0.25085170  0.0857175 -4.0787416 0.002436375 0.01949100 -1.279815
## Dendritic cells    0.37386979 -2.1849234  3.1619195 0.010692544 0.02738587 -2.638507
## CD14+ Monocytes   -0.10525402  1.2698117 -3.1226341 0.011413912 0.02738587 -2.709377
## B cells           -0.10478652  0.5516882 -3.0134349 0.013692935 0.02738587 -2.940542
## CD4 T cells       -0.07840101  2.0201947 -2.2318104 0.050869691 0.08139151 -4.128069
## FCGR3A+ Monocytes  0.07425165 -0.2567492  1.6647681 0.128337022 0.17111603 -4.935304
## NK cells           0.10270672  0.3797777  1.5181860 0.161321761 0.18436773 -5.247806
## Megakaryocytes     0.01377768 -1.8655172  0.1555131 0.879651456 0.87965146 -6.198336</code></pre>
<div id="multivariate-testing-along-a-tree" class="section level3">
<h3>Multivariate testing along a tree</h3>
<p>We can gain power by jointly testing multiple cell types using a
multivariate statistical model, instead of testing one cell type at a
time. Here we construct a hierarchical clustering between cell types
based on gene expression from pseudobulk, and perform a multivariate
test for each internal node of the tree based on its leaf nodes. The
results for the leaves are the same as from <code>topTable()</code>
above. At each internal node <code>treeTest()</code> performs a fixed
effects meta-analysis of the coefficients of the leaves while modeling
the covariance between coefficient estimates. In the backend, this is
implemented using <code>variancePartition::mvTest()</code> and <a
href="https://cran.r-project.org/package=remaCor">remaCor</a>
package.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># Perform multivariate test across the hierarchy</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">treeTest</span>(fit, cobj, hcl, <span class="at">coef =</span> <span class="st">&quot;StimStatusstim&quot;</span>)</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="co"># Plot hierarchy and testing results</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="fu">plotTreeTest</span>(res)</span></code></pre></div>
<p><img src="/Users/gabrielhoffman/workspace/repos/crumblr/docs/articles/crumblr_files/figure-html/treeTest-1.png" width="700" /></p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co"># Plot hierarchy and regression coefficients</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="fu">plotTreeTestBeta</span>(res)</span></code></pre></div>
<p><img src="/Users/gabrielhoffman/workspace/repos/crumblr/docs/articles/crumblr_files/figure-html/treeTest-2.png" width="700" /></p>
<div id="combined-plotting" class="section level4">
<h4>Combined plotting</h4>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="fu">plotTreeTestBeta</span>(res) <span class="sc">+</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;bottom&quot;</span>, <span class="at">legend.box =</span> <span class="st">&quot;vertical&quot;</span>) <span class="sc">|</span></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>  <span class="fu">plotForest</span>(res, <span class="at">hide =</span> <span class="cn">FALSE</span>) <span class="sc">|</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>  fig.vp</span></code></pre></div>
<p><img src="/Users/gabrielhoffman/workspace/repos/crumblr/docs/articles/crumblr_files/figure-html/combined-1.png" width="1152" /></p>
</div>
</div>
</div>
</div>
<div id="session-info" class="section level1">
<h1>Session Info</h1>
<details>
<pre><code>## R version 4.4.1 (2024-06-14)
## Platform: aarch64-apple-darwin23.5.0
## Running under: macOS Sonoma 14.7.1
## 
## Matrix products: default
## BLAS:   /Users/gabrielhoffman/prog/R-4.4.1/lib/libRblas.dylib 
## LAPACK: /opt/homebrew/Cellar/r/4.4.2_2/lib/R/lib/libRlapack.dylib;  LAPACK version 3.12.0
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## time zone: America/New_York
## tzcode source: internal
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] variancePartition_1.36.3 BiocParallel_1.40.0      limma_3.62.1            
## [4] crumblr_0.99.14          ggplot2_3.5.1           
## 
## loaded via a namespace (and not attached):
##   [1] Rdpack_2.6.2                bitops_1.0-9                gridExtra_2.3              
##   [4] rlang_1.1.4                 magrittr_2.0.3              matrixStats_1.4.1          
##   [7] compiler_4.4.1              reshape2_1.4.4              systemfonts_1.1.0          
##  [10] vctrs_0.6.5                 RcppZiggurat_0.1.6          stringr_1.5.1              
##  [13] pkgconfig_2.0.3             crayon_1.5.3                fastmap_1.2.0              
##  [16] backports_1.5.0             XVector_0.46.0              labeling_0.4.3             
##  [19] caTools_1.18.3              utf8_1.2.4                  rmarkdown_2.29             
##  [22] nloptr_2.1.1                UCSC.utils_1.2.0            ragg_1.3.3                 
##  [25] purrr_1.0.2                 xfun_0.49                   Rfast_2.1.0                
##  [28] zlibbioc_1.52.0             cachem_1.1.0                aplot_0.2.4                
##  [31] GenomeInfoDb_1.42.1         jsonlite_1.8.9              EnvStats_3.0.0             
##  [34] remaCor_0.0.18              DelayedArray_0.32.0         broom_1.0.7                
##  [37] parallel_4.4.1              R6_2.5.1                    stringi_1.8.4              
##  [40] bslib_0.8.0                 boot_1.3-31                 numDeriv_2016.8-1.1        
##  [43] GenomicRanges_1.58.0        jquerylib_0.1.4             Rcpp_1.0.13-1              
##  [46] SummarizedExperiment_1.36.0 iterators_1.0.14            knitr_1.49                 
##  [49] IRanges_2.40.1              Matrix_1.7-1                splines_4.4.1              
##  [52] tidyselect_1.2.1            viridis_0.6.5               abind_1.4-8                
##  [55] yaml_2.3.10                 gplots_3.2.0                codetools_0.2-20           
##  [58] plyr_1.8.9                  lmerTest_3.1-3              lattice_0.22-6             
##  [61] tibble_3.2.1                Biobase_2.66.0              treeio_1.26.0              
##  [64] withr_3.0.2                 evaluate_1.0.1              gridGraphics_0.5-1         
##  [67] desc_1.4.3                  RcppParallel_5.1.9          pillar_1.9.0               
##  [70] ggtree_3.10.1               MatrixGenerics_1.18.0       KernSmooth_2.23-24         
##  [73] stats4_4.4.1                ggfun_0.1.8                 generics_0.1.3             
##  [76] S4Vectors_0.44.0            munsell_0.5.1               scales_1.3.0               
##  [79] aod_1.3.3                   tidytree_0.4.6              minqa_1.2.8                
##  [82] gtools_3.9.5                RhpcBLASctl_0.23-42         glue_1.8.0                 
##  [85] lazyeval_0.2.2              tools_4.4.1                 fANCOVA_0.6-1              
##  [88] lme4_1.1-35.5               mvtnorm_1.3-2               fs_1.6.5                   
##  [91] grid_4.4.1                  tidyr_1.3.1                 ape_5.8-1                  
##  [94] rbibutils_2.3               colorspace_2.1-1            SingleCellExperiment_1.28.1
##  [97] nlme_3.1-166                GenomeInfoDbData_1.2.13     patchwork_1.3.0            
## [100] cli_3.6.3                   textshaping_0.4.1           fansi_1.0.6                
## [103] viridisLite_0.4.2           S4Arrays_1.6.0              dplyr_1.1.4                
## [106] corpcor_1.6.10              gtable_0.3.6                yulab.utils_0.1.8          
## [109] sass_0.4.9                  digest_0.6.37               BiocGenerics_0.52.0        
## [112] pbkrtest_0.5.3              SparseArray_1.6.0           ggplotify_0.1.2            
## [115] htmlwidgets_1.6.4           farver_2.1.2                htmltools_0.5.8.1          
## [118] pkgdown_2.1.1               lifecycle_1.0.4             httr_1.4.7                 
## [121] statmod_1.5.0               MASS_7.3-61</code></pre>
</details>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>
