<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Integration with dreamlet / SingleCellExperiment • crumblr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous">
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Integration with dreamlet / SingleCellExperiment">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">crumblr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.22</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/crumblr.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/crumblr_theory.html">Normal approximation vs. empirical simulation</a></li>
    <li><a class="dropdown-item" href="../articles/integration.html">Integration with dreamlet / SingleCellExperiment</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/DiseaseNeurogenomics/crumblr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Integration with dreamlet / SingleCellExperiment</h1>
                        <h4 data-toc-skip class="author">Developed by <a href="http://gabrielhoffman.github.io/" class="external-link">Gabriel Hoffman</a>
</h4>
            
            <h4 data-toc-skip class="date">Run on 2025-08-11
14:19:51.385562</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/DiseaseNeurogenomics/crumblr/blob/HEAD/vignettes/integration.Rmd" class="external-link"><code>vignettes/integration.Rmd</code></a></small>
      <div class="d-none name"><code>integration.Rmd</code></div>
    </div>

    
    
<style>
body {
text-align: justify}
</style>
<div class="section level2">
<h2 id="load-and-process-single-cell-data">Load and process single cell data<a class="anchor" aria-label="anchor" href="#load-and-process-single-cell-data"></a>
</h2>
<p>Here we perform analysis of PBMCs from 8 individuals stimulated with
interferon-β <a href="https://www.nature.com/articles/nbt.4042" class="external-link">Kang, et
al, 2018, Nature Biotech</a>. We perform standard processing with <a href="https://gabrielhoffman.github.io/dreamlet/index.html" class="external-link">dreamlet</a>
to compute pseudobulk before applying <code>crumblr</code>.</p>
<p>Here, single cell RNA-seq data is downloaded from <a href="https://bioconductor.org/packages/ExperimentHub/" class="external-link">ExperimentHub</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://DiseaseNeurogenomics.github.io/dreamlet" class="external-link">dreamlet</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HelenaLC/muscat" class="external-link">muscat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/ExperimentHub" class="external-link">ExperimentHub</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Download data, specifying EH2259 for the Kang, et al. study</span></span>
<span><span class="va">eh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ExperimentHub/man/ExperimentHub-class.html" class="external-link">ExperimentHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH2259"</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">ind</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">ind</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># only keep singlet cells with sufficient reads</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="op">]</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">multiplets</span> <span class="op">==</span> <span class="st">"singlet"</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># compute QC metrics</span></span>
<span><span class="va">qc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html" class="external-link">perCellQCMetrics</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># remove cells with few or many detected genes</span></span>
<span><span class="va">ol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span>metric <span class="op">=</span> <span class="va">qc</span><span class="op">$</span><span class="va">detected</span>, nmads <span class="op">=</span> <span class="fl">2</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="op">!</span><span class="va">ol</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># set variable indicating stimulated (stim) or control (ctrl)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">StimStatus</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">$</span><span class="va">stim</span></span></code></pre></div>
<div class="section level3">
<h3 id="aggregate-to-pseudobulk">Aggregate to pseudobulk<a class="anchor" aria-label="anchor" href="#aggregate-to-pseudobulk"></a>
</h3>
<p>Dreamlet creates the pseudobulk dataset:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Since 'ind' is the individual and 'StimStatus' is the stimulus status,</span></span>
<span><span class="co"># create unique identifier for each sample</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">StimStatus</span>, <span class="va">sce</span><span class="op">$</span><span class="va">ind</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create pseudobulk data by specifying cluster_id and sample_id for aggregating cells</span></span>
<span><span class="va">pb</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://DiseaseNeurogenomics.github.io/dreamlet/reference/aggregateToPseudoBulk.html" class="external-link">aggregateToPseudoBulk</a></span><span class="op">(</span><span class="va">sce</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  cluster_id <span class="op">=</span> <span class="st">"cell"</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="st">"id"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="process-data">Process data<a class="anchor" aria-label="anchor" href="#process-data"></a>
</h3>
<p>Here we evaluate whether the observed cell proportions change in
response to interferon-β.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://DiseaseNeurogenomics.github.io/crumblr" class="external-link">crumblr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># use dreamlet::cellCounts() to extract data</span></span>
<span><span class="fu"><a href="http://DiseaseNeurogenomics.github.io/dreamlet/reference/cellCounts.html" class="external-link">cellCounts</a></span><span class="op">(</span><span class="va">pb</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##          B cells CD14+ Monocytes CD4 T cells</span></span>
<span><span class="co">## ctrl101      101             136         288</span></span>
<span><span class="co">## ctrl1015     424             644         819</span></span>
<span><span class="co">## ctrl1016     119             315         413</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Apply crumblr transformation</span></span>
<span><span class="co"># cobj is an EList object compatable with limma workflow</span></span>
<span><span class="co"># cobj$E stores transformed values</span></span>
<span><span class="co"># cobj$weights stores precision weights</span></span>
<span><span class="va">cobj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/crumblr.html">crumblr</a></span><span class="op">(</span><span class="fu"><a href="http://DiseaseNeurogenomics.github.io/dreamlet/reference/cellCounts.html" class="external-link">cellCounts</a></span><span class="op">(</span><span class="va">pb</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="analysis">Analysis<a class="anchor" aria-label="anchor" href="#analysis"></a>
</h3>
<p>Now continue on with the downstream analysis</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/variancePartition" class="external-link">variancePartition</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/dream-method.html" class="external-link">dream</a></span><span class="op">(</span><span class="va">cobj</span>, <span class="op">~</span> <span class="va">StimStatus</span> <span class="op">+</span> <span class="va">ind</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">pb</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">topTable</span><span class="op">(</span><span class="va">fit</span>, coef <span class="op">=</span> <span class="st">"StimStatusstim"</span>, number <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                         logFC    AveExpr          t     P.Value  adj.P.Val         B</span></span>
<span><span class="co">## CD8 T cells       -0.25085170  0.0857175 -4.0787416 0.002436375 0.01949100 -1.279815</span></span>
<span><span class="co">## Dendritic cells    0.37386979 -2.1849234  3.1619195 0.010692544 0.02738587 -2.638507</span></span>
<span><span class="co">## CD14+ Monocytes   -0.10525402  1.2698117 -3.1226341 0.011413912 0.02738587 -2.709377</span></span>
<span><span class="co">## B cells           -0.10478652  0.5516882 -3.0134349 0.013692935 0.02738587 -2.940542</span></span>
<span><span class="co">## CD4 T cells       -0.07840101  2.0201947 -2.2318104 0.050869691 0.08139151 -4.128069</span></span>
<span><span class="co">## FCGR3A+ Monocytes  0.07425165 -0.2567492  1.6647681 0.128337022 0.17111603 -4.935304</span></span>
<span><span class="co">## NK cells           0.10270672  0.3797777  1.5181860 0.161321761 0.18436773 -5.247806</span></span>
<span><span class="co">## Megakaryocytes     0.01377768 -1.8655172  0.1555131 0.879651456 0.87965146 -6.198336</span></span></code></pre>
<p>Given the results here, we see that CD8 T cells at others change
relative abundance following treatment with interferon-β.</p>
</div>
<div class="section level3">
<h3 id="multivariate-testing-along-a-tree">Multivariate testing along a tree<a class="anchor" aria-label="anchor" href="#multivariate-testing-along-a-tree"></a>
</h3>
<p>ere we construct a hierarchical clustering between cell types based
on gene expression from pseudobulk, and perform a multivariate test for
each internal node of the tree based on its leaf nodes. The results for
the leaves are the same as from <code>topTable()</code> above.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># hierarchical cluster based on pseudobulked gene expression</span></span>
<span><span class="va">hcl</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://DiseaseNeurogenomics.github.io/dreamlet/reference/buildClusterTreeFromPB.html" class="external-link">buildClusterTreeFromPB</a></span><span class="op">(</span><span class="va">pb</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform multivariate test across the hierarchy</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/treeTest.html">treeTest</a></span><span class="op">(</span><span class="va">fit</span>, <span class="va">cobj</span>, <span class="va">hcl</span>, coef <span class="op">=</span> <span class="st">"StimStatusstim"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot hierarchy and testing results</span></span>
<span><span class="fu"><a href="../reference/plotTreeTest.html">plotTreeTest</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<p><img src="integration_files/figure-html/tree-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<details><pre><code><span><span class="co">## R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin23.6.0</span></span>
<span><span class="co">## Running under: macOS Sonoma 14.7.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS/LAPACK: /opt/homebrew/Cellar/openblas/0.3.30/lib/libopenblasp-r0.3.30.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/New_York</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] crumblr_0.99.22             muscData_1.22.0             scater_1.36.0              </span></span>
<span><span class="co">##  [4] scuttle_1.18.0              ExperimentHub_2.16.1        AnnotationHub_3.16.1       </span></span>
<span><span class="co">##  [7] BiocFileCache_2.16.1        dbplyr_2.5.0                muscat_1.22.0              </span></span>
<span><span class="co">## [10] dreamlet_1.6.0              SingleCellExperiment_1.30.1 SummarizedExperiment_1.38.1</span></span>
<span><span class="co">## [13] Biobase_2.68.0              GenomicRanges_1.60.0        GenomeInfoDb_1.44.1        </span></span>
<span><span class="co">## [16] IRanges_2.42.0              S4Vectors_0.46.0            BiocGenerics_0.54.0        </span></span>
<span><span class="co">## [19] generics_0.1.4              MatrixGenerics_1.20.0       matrixStats_1.5.0          </span></span>
<span><span class="co">## [22] variancePartition_1.37.4    BiocParallel_1.42.1         limma_3.64.3               </span></span>
<span><span class="co">## [25] ggplot2_3.5.2               BiocStyle_2.36.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] fs_1.6.6                  bitops_1.0-9              httr_1.4.7               </span></span>
<span><span class="co">##   [4] RColorBrewer_1.1-3        doParallel_1.0.17         Rgraphviz_2.52.0         </span></span>
<span><span class="co">##   [7] numDeriv_2016.8-1.1       tools_4.5.1               sctransform_0.4.2        </span></span>
<span><span class="co">##  [10] backports_1.5.0           R6_2.6.1                  metafor_4.8-0            </span></span>
<span><span class="co">##  [13] lazyeval_0.2.2            mgcv_1.9-3                GetoptLong_1.0.5         </span></span>
<span><span class="co">##  [16] withr_3.0.2               prettyunits_1.2.0         gridExtra_2.3            </span></span>
<span><span class="co">##  [19] cli_3.6.5                 textshaping_1.0.1         labeling_0.4.3           </span></span>
<span><span class="co">##  [22] sass_0.4.10               KEGGgraph_1.68.0          SQUAREM_2021.1           </span></span>
<span><span class="co">##  [25] mvtnorm_1.3-3             blme_1.0-6                pkgdown_2.1.3            </span></span>
<span><span class="co">##  [28] mixsqp_0.3-54             yulab.utils_0.2.0         systemfonts_1.2.3        </span></span>
<span><span class="co">##  [31] zenith_1.10.0             parallelly_1.45.1         invgamma_1.2             </span></span>
<span><span class="co">##  [34] RSQLite_2.4.2             gridGraphics_0.5-1        shape_1.4.6.1            </span></span>
<span><span class="co">##  [37] gtools_3.9.5              dplyr_1.1.4               Matrix_1.7-3             </span></span>
<span><span class="co">##  [40] metadat_1.4-0             ggbeeswarm_0.7.2          abind_1.4-8              </span></span>
<span><span class="co">##  [43] lifecycle_1.0.4           yaml_2.3.10               edgeR_4.6.3              </span></span>
<span><span class="co">##  [46] mathjaxr_1.8-0            gplots_3.2.0              SparseArray_1.8.1        </span></span>
<span><span class="co">##  [49] grid_4.5.1                blob_1.2.4                crayon_1.5.3             </span></span>
<span><span class="co">##  [52] lattice_0.22-7            beachmat_2.24.0           msigdbr_25.1.1           </span></span>
<span><span class="co">##  [55] annotate_1.86.1           KEGGREST_1.48.1           pillar_1.11.0            </span></span>
<span><span class="co">##  [58] knitr_1.50                ComplexHeatmap_2.24.1     rjson_0.2.23             </span></span>
<span><span class="co">##  [61] boot_1.3-31               corpcor_1.6.10            future.apply_1.20.0      </span></span>
<span><span class="co">##  [64] codetools_0.2-20          glue_1.8.0                ggfun_0.2.0              </span></span>
<span><span class="co">##  [67] data.table_1.17.8         treeio_1.32.0             vctrs_0.6.5              </span></span>
<span><span class="co">##  [70] png_0.1-8                 Rdpack_2.6.4              gtable_0.3.6             </span></span>
<span><span class="co">##  [73] assertthat_0.2.1          cachem_1.1.0              zigg_0.0.2               </span></span>
<span><span class="co">##  [76] xfun_0.52                 mime_0.13                 rbibutils_2.3            </span></span>
<span><span class="co">##  [79] S4Arrays_1.8.1            Rfast_2.1.5.1             reformulas_0.4.1         </span></span>
<span><span class="co">##  [82] iterators_1.0.14          statmod_1.5.0             nlme_3.1-168             </span></span>
<span><span class="co">##  [85] pbkrtest_0.5.5            ggtree_3.16.3             bit64_4.6.0-1            </span></span>
<span><span class="co">##  [88] filelock_1.0.3            progress_1.2.3            EnvStats_3.1.0           </span></span>
<span><span class="co">##  [91] bslib_0.9.0               TMB_1.9.17                irlba_2.3.5.1            </span></span>
<span><span class="co">##  [94] vipor_0.4.7               KernSmooth_2.23-26        colorspace_2.1-1         </span></span>
<span><span class="co">##  [97] rmeta_3.0                 DBI_1.2.3                 DESeq2_1.48.1            </span></span>
<span><span class="co">## [100] tidyselect_1.2.1          bit_4.6.0                 compiler_4.5.1           </span></span>
<span><span class="co">## [103] curl_6.4.0                graph_1.86.0              BiocNeighbors_2.2.0      </span></span>
<span><span class="co">## [106] desc_1.4.3                DelayedArray_0.34.1       bookdown_0.43            </span></span>
<span><span class="co">## [109] scales_1.4.0              caTools_1.18.3            remaCor_0.0.18           </span></span>
<span><span class="co">## [112] rappdirs_0.3.3            stringr_1.5.1             digest_0.6.37            </span></span>
<span><span class="co">## [115] minqa_1.2.8               rmarkdown_2.29            aod_1.3.3                </span></span>
<span><span class="co">## [118] XVector_0.48.0            RhpcBLASctl_0.23-42       htmltools_0.5.8.1        </span></span>
<span><span class="co">## [121] pkgconfig_2.0.3           lme4_1.1-37               sparseMatrixStats_1.20.0 </span></span>
<span><span class="co">## [124] mashr_0.2.79              fastmap_1.2.0             rlang_1.1.6              </span></span>
<span><span class="co">## [127] GlobalOptions_0.1.2       htmlwidgets_1.6.4         UCSC.utils_1.4.0         </span></span>
<span><span class="co">## [130] DelayedMatrixStats_1.30.0 farver_2.1.2              jquerylib_0.1.4          </span></span>
<span><span class="co">## [133] jsonlite_2.0.0            BiocSingular_1.24.0       RCurl_1.98-1.17          </span></span>
<span><span class="co">## [136] magrittr_2.0.3            ggplotify_0.1.2           GenomeInfoDbData_1.2.14  </span></span>
<span><span class="co">## [139] patchwork_1.3.1           Rcpp_1.1.0                ape_5.8-1                </span></span>
<span><span class="co">## [142] babelgene_22.9            viridis_0.6.5             EnrichmentBrowser_2.38.0 </span></span>
<span><span class="co">## [145] stringi_1.8.7             MASS_7.3-65               plyr_1.8.9               </span></span>
<span><span class="co">## [148] listenv_0.9.1             parallel_4.5.1            ggrepel_0.9.6            </span></span>
<span><span class="co">## [151] Biostrings_2.76.0         splines_4.5.1             hms_1.1.3                </span></span>
<span><span class="co">## [154] circlize_0.4.16           locfit_1.5-9.12           reshape2_1.4.4           </span></span>
<span><span class="co">## [157] ScaledMatrix_1.16.0       BiocVersion_3.21.1        XML_3.99-0.18            </span></span>
<span><span class="co">## [160] evaluate_1.0.4            RcppParallel_5.1.10.9000  BiocManager_1.30.26      </span></span>
<span><span class="co">## [163] nloptr_2.2.1              foreach_1.5.2             tidyr_1.3.1              </span></span>
<span><span class="co">## [166] purrr_1.1.0               future_1.67.0             clue_0.3-66              </span></span>
<span><span class="co">## [169] scattermore_1.2           ashr_2.2-63               rsvd_1.0.5               </span></span>
<span><span class="co">## [172] broom_1.0.9               xtable_1.8-4              tidytree_0.4.6           </span></span>
<span><span class="co">## [175] fANCOVA_0.6-1             viridisLite_0.4.2         ragg_1.4.0               </span></span>
<span><span class="co">## [178] truncnorm_1.0-9           tibble_3.3.0              aplot_0.2.8              </span></span>
<span><span class="co">## [181] lmerTest_3.1-3            glmmTMB_1.1.11            memoise_2.0.1            </span></span>
<span><span class="co">## [184] beeswarm_0.4.0            AnnotationDbi_1.70.0      cluster_2.1.8.1          </span></span>
<span><span class="co">## [187] globals_0.18.0            GSEABase_1.70.0</span></span></code></pre>
</details>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
